text/0000-custom-dst.md

```diff
@@ -0,0 +1,479 @@
+- Feature Name: `custom_dst`
+- Start Date: 2018-11-06
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Allow Rust code to define dynamically sized types with custom thick (and thin)
+pointers, and define slice functions in terms of these, instead of transmute.
+Also, convert the `CStr` type to use this functionality,
+and make it a thin pointer; this will allow use with FFI.
+
+# Motivation
+[motivation]: #motivation
+
+As of right now, the lack of custom DSTs in Rust means that we can't communicate
+with C in important ways - we lack the ability to define a `CStr` in the
+language such that `&CStr` is compatible with `char const *`,
+and we lack the ability to communicate nicely with C code that uses
+[Flexible Array Members](https://en.wikipedia.org/wiki/Flexible_array_member).
+This RFC attempts to fix this,
+as well as introduce more correctness to existing practices.
+
+Apart from FFI, it also has usecases for indexing and slicing 2-d arrays.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+There's a new language trait in the standard library, similar to `Sized`,
+under `std::marker`:
+
+```rust
+unsafe trait DynamicallySized {
```

Comment:
One weird quirk: `&dyn DynamicallySized` is nonsense because the metadata has unknown size, so we can't have a reference containing the metadata. Clearly the bound here to express that isn't : Sized. What is it, exactly? Is the fact that this bound is inexpressible just a quirk that we paper over via compiler builtins? cc @eddyb who may have thought about this before.

Reply:
> because the metadata has unknown size

This is the same as `&dyn Iterator` is nonsense, right? You need `&dyn DynSized<Meta=KnownStuff>`. 

The actual representation (`(&(), KnownStuff, *VTable)` or forbidden) is another matter.

Reply:
Ah, good catch. Thanks, you're right of course.